### ISSUE TITLE
Root certificate doesn't work on Catalina

### ISSUE THREAD

**Issue by dutzi** [mitmproxy/mitmproxy#3682]
##### Steps to reproduce the problem:

1. Run mitmproxy
2. Visit mitm.it
3. Download the certificate for macOs
4. Add it (tried both Login and System keychains)

![image](https://user-images.githubusercontent.com/927310/67876927-5f30f200-fb41-11e9-99d6-d16bbc63f82e.png)

But when visiting a secured website the site doesn't load.

This started happening after I upgraded to macOs Catalina

**Comment by Kriechi** [comment]
Are you running the most recent mitmproxy version?
We recently merged a PR into git master which should fix a similar issue for iOS 13 https://github.com/mitmproxy/mitmproxy/pull/3659
Would you mind trying it for macOS?

**Comment by dutzi** [comment]
Still does not work, I get the following error when trying to visit https://example.com

```
NET::ERR_CERT_WEAK_KEY
Subject: www.example.org

Issuer: mitmproxy

Expires on: Nov 2, 2021

Current date: Nov 3, 2019

PEM encoded chain:
-----BEGIN CERTIFICATE-----
MIICRTCCAa6gAwIBAgIGDk30j8CJMA0GCSqGSIb3DQEBCwUAMCgxEjAQBgNVBAMM
CW1pdG1wcm94eTESMBAGA1UECgwJbWl0bXByb3h5MB4XDTE5MTEwMTE2MTcxM1oX
DTIxMTEwMjE2MTcxM1owGjEYMBYGA1UEAwwPd3d3LmV4YW1wbGUub3JnMIGfMA0G
CSqGSIb3DQEBAQUAA4GNADCBiQKBgQC528bbuqNz/M6eNIAmJ+xf8ss57f4zl/pL
zVCBD6l7P/buh8AbHih0zsYGv4pGbf9+aGivJCEyyJSvfkI8WowtR1pNw79mcBwv
xqpRTeYFEoiIBmZ4MZ72eUFS3RGibtvQKaGhZhCP6JZhIxeRYLmq6lqSszf1eL4O
eknr9d6U5QIDAQABo4GHMIGEMIGBBgNVHREEejB4ggtleGFtcGxlLm5ldIILZXhh
bXBsZS5lZHWCD3d3dy5leGFtcGxlLmNvbYIPd3d3LmV4YW1wbGUub3Jngg93d3cu
ZXhhbXBsZS5uZXSCD3d3dy5leGFtcGxlLmVkdYILZXhhbXBsZS5jb22CC2V4YW1w
bGUub3JnMA0GCSqGSIb3DQEBCwUAA4GBAH6pGkPWGwivWN4Ca6EO4WIi1mebyI7R
O6P6SED5RO+HWzWyT/mcNfS9WP44kzNAHbkgRvD1UWP/DIUzFnL8k2VBI6BQZAjT
tUSwH/U3/DQY/khGCQvJlf2A7nVxJFEfl6Pse9gyj2X7rGHGO77QGrEJcJ5Ic4tZ
kklhf7DIv4VJ
-----END CERTIFICATE-----
-----BEGIN CERTIFICATE-----
MIICnDCCAgWgAwIBAgIGDjQr7RhmMA0GCSqGSIb3DQEBBQUAMCgxEjAQBgNVBAMM
CW1pdG1wcm94eTESMBAGA1UECgwJbWl0bXByb3h5MB4XDTE5MDYyNjEyMTAwNFoX
DTIxMDYxNzEyMTAwNFowKDESMBAGA1UEAwwJbWl0bXByb3h5MRIwEAYDVQQKDAlt
aXRtcHJveHkwgZ8wDQYJKoZIhvcNAQEBBQADgY0AMIGJAoGBALnbxtu6o3P8zp40
gCYn7F/yyznt/jOX+kvNUIEPqXs/9u6HwBseKHTOxga/ikZt/35oaK8kITLIlK9+
QjxajC1HWk3Dv2ZwHC/GqlFN5gUSiIgGZngxnvZ5QVLdEaJu29ApoaFmEI/olmEj
F5FguarqWpKzN/V4vg56Sev13pTlAgMBAAGjgdAwgc0wDwYDVR0TAQH/BAUwAwEB
/zARBglghkgBhvhCAQEEBAMCAgQweAYDVR0lBHEwbwYIKwYBBQUHAwEGCCsGAQUF
BwMCBggrBgEFBQcDBAYIKwYBBQUHAwgGCisGAQQBgjcCARUGCisGAQQBgjcCARYG
CisGAQQBgjcKAwEGCisGAQQBgjcKAwMGCisGAQQBgjcKAwQGCWCGSAGG+EIEATAO
BgNVHQ8BAf8EBAMCAQYwHQYDVR0OBBYEFJq3hb5mLxsqAZ43IVEKiX7kw6Q1MA0G
CSqGSIb3DQEBBQUAA4GBAEeDwrUoAcfd6kFh4INI1EF/dxyBGv3NNbSIWjQgqpeh
vMZ5Vk6FNDDvZ0aw6i4LoTO885Q0AejljOW9Xvz8KTy4ven0f49fM6GUewh5G2zH
YZQ3Oyq390X9fl7m0p1pTVEwYSHB9mOua92Mehvszv4oXgYjw4+nobJx1jrI0vpy
-----END CERTIFICATE-----

```

**Comment by vin01** [comment]
@dutzi The certificate chain you have shared indicates you are using `1024` bit keys for certificates, are you overriding the default value of 2048?  You might want to double check when you are using the commit specified by Kriechi if you are actually using the newer version

This was updated long time back to use 2048 bit key length from what I can see: https://github.com/mitmproxy/mitmproxy/commit/041ca5c499369ffbf115e4451b85aee77e3095c0

2048 and above key lengths shall work, 1024 will not.

https://community.sophos.com/products/unified-threat-management/f/web-protection-web-filtering-application-visibility-control/115345/issue-with-ios-13-mac-os-10-15-ssl-certificate-requirments-for-transparent-proxy

**Comment by dutzi** [comment]
Hey @vin01 that's the certificate I'm getting from mitm.it using

```
Mitmproxy: 4.0.4
Python:    3.7.5
OpenSSL:   OpenSSL 1.1.1d  10 Sep 2019
Platform:  Darwin-19.0.0-x86_64-i386-64bit
```

I think it includes the commit as it is very old.

Is there a cache where mitmproxy is pulling the certificate from? Maybe it's using an old one?

Does this screenshot of the cert's details help?

![image](https://user-images.githubusercontent.com/927310/68606167-86c77900-04b6-11ea-88cc-b35ea23035c1.png)

**Comment by dutzi** [comment]
I ran `find / | grep netlib/certutils.py` and found only:

```
/usr/local/lib/python2.7/site-packages/netlib/certutils.py
/usr/local/lib/python2.7/site-packages/netlib/certutils.pyc
/System/Volumes/Data/usr/local/lib/python2.7/site-packages/netlib/certutils.py
/System/Volumes/Data/usr/local/lib/python2.7/site-packages/netlib/certutils.pyc
```

Shouldn't I have found it under python3.X?

**Comment by Kriechi** [comment]
`netlib` is a very old dependency of mitmproxy. In recent version it is no longer used.
You can uninstall netlib without problems.

The mitmproxy CA files should be in `~/.mitmproxy`. If you delete them, the next time you start mitmproxy will re-create them with the most recent settings/options.

**Comment by dutzi** [comment]
@Kriechi Thanks. I deleted `~/.mitmproxy` and re-ran mitmdump, it created the cert files, I installed them through mitm.it but I still can't load https sites, Chrome shows me this warning:

![image](https://user-images.githubusercontent.com/927310/68754684-d8cfe200-060f-11ea-807a-33f84e80fb29.png)

The certificate is now 2048 bit SHA-256:

![image](https://user-images.githubusercontent.com/927310/68754748-f604b080-060f-11ea-9b80-e6d0fefd779c.png)

**Comment by dutzi** [comment]
Sometimes when browsing to secure sites Chrome would open a new tab to http://www.gstatic.com/generate_204, which shows an empty page.

**Comment by dutzi** [comment]
plz help

**Comment by roipeker** [comment]
Experiencing exactly the same issue as you do, on macOS Catalina as "proxy" and every device (iOS, Android, etc)

**Comment by mhils** [comment]
I believe this is fixed with the 5.0 release. 😃
